substrate/frame/election-provider-multi-block/src/lib.rs

```diff
@@ -0,0 +1,1193 @@
+				Phase::Signed
+					if remaining_blocks <= signed_validation_deadline &&
+						remaining_blocks > unsigned_deadline =>
+				{
+					Self::phase_transition(Phase::SignedValidation(now));
+					T::WeightInfo::on_phase_transition()
+				},
+
+				// start unsigned phase. The `unsigned` sub-pallet will take further actions now.
+				Phase::Signed | Phase::SignedValidation(_) | Phase::Snapshot(0)
+					if remaining_blocks <= unsigned_deadline && remaining_blocks > Zero::zero() =>
+				{
+					Self::phase_transition(Phase::Unsigned(now));
+					T::WeightInfo::on_phase_transition()
+				},
+
+				// start export phase.
+				Phase::Unsigned(_) if now == next_election.saturating_sub(export_deadline) => {
+					Self::phase_transition(Phase::Export(now));
+					T::WeightInfo::on_phase_transition()
+				},
+
+				// election solution **MAY** be ready, start export phase to allow external pallets
+                // to request paged election solutions.
+				Phase::Export(started_at) => Self::do_export_phase(now, started_at),
+
+				_ => T::WeightInfo::on_initialize_do_nothing(),
+			}
+		}
+
+		fn integrity_test() {
+			// the signed validator phase must not be less than the number of pages of a
+			// submission.
+			assert!(
+				T::SignedValidationPhase::get() <= T::Pages::get().into(),
+				"signed validaton phase ({:?}) should not be less than the number of pages per submission ({:?})",
+				T::SignedValidationPhase::get(),
+				T::Pages::get(),
+			);
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+}
+
+/// Wrapper struct for working with snapshots.
+///
+/// It manages the following storage items:
+///
+/// - [`PagedVoterSnapshot`]: Paginated map of voters.
+/// - [`TargetSnapshot`]: Single page, bounded list of targets.
+///
+/// To ensure correctness and data consistency, all the reads and writes to storage items related
+/// to the snapshot and "wrapped" by this struct must be performed through the methods exposed by
+/// the implementation of [`Snapshot`].
+pub(crate) struct Snapshot<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> Snapshot<T> {
+	/// Returns the targets snapshot.
+	///
+	/// The target snapshot is single paged.
+	fn targets() -> Option<TargetPageOf<T>> {
+		TargetSnapshot::<T>::get()
+	}
+
+	/// Sets a page of targets in the snapshot's storage.
+	///
+	/// The target snapshot is single paged.
+	fn set_targets(targets: TargetPageOf<T>) {
+		TargetSnapshot::<T>::set(Some(targets));
+	}
+
+	/// Returns whether the target snapshot exists in storage.
+	fn targets_snapshot_exists() -> bool {
+		TargetSnapshot::<T>::get().is_some()
+	}
+
+	/// Returns the number of desired targets, as defined by [`T::DataProvider`].
+	fn desired_targets() -> Option<u32> {
+		match T::DataProvider::desired_targets() {
+			Ok(desired) => Some(desired),
+			Err(err) => {
+				defensive!(
+					"error fetching the desired targets from the election data provider {}",
+					err
+				);
+				None
```

Comment:
Will "None" be ever returned? I wrote a test and it panicks at "defensive" already